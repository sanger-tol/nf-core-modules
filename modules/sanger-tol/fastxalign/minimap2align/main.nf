process FASTXALIGN_MINIMAP2ALIGN {
    tag "$meta.id"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/63/637ecb387eaafe0b1689e3e32c5eda589e016cfd46c482946425181f69f0733e/data' :
        'community.wave.seqera.io/library/htslib_minimap2_pyfastx_samtools_click:bfd8f60cc27aa6d6' }"

    input:
    tuple val(meta), path(fastx), path(fxi), val(chunkn), val(range), path(reference)
    val bam_format

    output:
    tuple val(meta), path("*.bam")   , emit: bam, optional: true
    tuple val(meta), path("*.paf.gz"), emit: paf, optional: true
    path "versions.yml"              , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    // WARNING: This module includes the slice_fasta.py script as a module binary in
    // ${moduleDir}/resources/usr/bin/slice_fasta.py. To use this module, you will
    // either have to copy this file to ${projectDir}/bin or set the option
    // nextflow.enable.moduleBinaries = true
    // in your nextflow.config file.
    def args1       = task.ext.args1  ?: ''
    def args2       = task.ext.args2  ?: ''
    def args3       = task.ext.args3  ?: ''
    def prefix      = task.ext.prefix ?: "${fastx}.${chunkn}.${meta.id}"
    def post_filter = args2 ? "samtools view -h ${args2} - |" : ''
    def sort_bam    = "samtools sort -@ ${task.cpus > 1 ? task.cpus - 1 : 1} -o ${prefix}.bam -T ${prefix}_sort_tmp ${args3} -"
    def bam_output  = bam_format      ? "-a | ${post_filter} ${sort_bam}" : "| bgzip -@ ${task.cpus} > ${prefix}.paf.gz"
    """
    slice_fasta.py slice ${fastx} ${range[0]} ${range[1]} | \\
        minimap2 -t${task.cpus} ${args1} ${reference} - \\
        ${bam_output}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        slice_fasta.py: \$(slice_fasta.py --version)
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' )
        minimap2: \$(minimap2 --version | sed 's/minimap2 //g')
    END_VERSIONS
    """

    stub:
    def prefix  = task.ext.prefix ?: "${fastx}.${chunkn}.${meta.id}"
    """
    touch ${prefix}.bam
    echo "" | gzip > ${prefix}.paf.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        slice_fasta.py: \$(slice_fasta.py --version)
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' )
        minimap2: \$(minimap2 --version | sed 's/minimap2 //g')
    END_VERSIONS
    """
}
