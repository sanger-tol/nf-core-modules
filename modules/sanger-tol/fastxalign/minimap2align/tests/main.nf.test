nextflow_process {

    name "Test Process FASTXALIGN_MINIMAP2ALIGN"
    script "../main.nf"
    process "FASTXALIGN_MINIMAP2ALIGN"

    tag "modules"
    tag "modules_sangertol"
    tag "fastxalign"
    tag "fastxalign/pyfastxindexfasta"
    tag "fastxalign/minimap2align"

    config "./nextflow.config"

    setup {
        run("FASTXALIGN_PYFASTXINDEXFASTA") {
            script "../../pyfastxindexfasta/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id: "test" ],
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/genomic_data/baUndUnlc1/pacbio/fasta/HiFi.reads.fasta', checkIfExists: true)
                ])
                """
            }
        }
    }

    test("Undibacterium_unclassified - fasta - bam out - no filter") {

        when {

            params {
                minimap2_args      = "-x map-hifi"
                samtools_view_args = ""
                samtools_sort_args = "--write-index -l1"
            }

            process {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/genomic_data/baUndUnlc1/pacbio/fasta/HiFi.reads.fasta', checkIfExists: true),
                    0,
                    [0, 500],
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/assembly/draft/baUndUnlc1.hic.hap1.p_ctg.fa.gz', checkIfExists: true)
                ]).combine(FASTXALIGN_PYFASTXINDEXFASTA.out.index, by: 0)
                .map { meta, reads, chunkn, range, reference, index, count ->
                    [ meta, reads, index, chunkn, range, reference ]
                }
                input[1] = true
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    bam(process.out.bam.get(0).get(1)).getStatistics(),
                    process.out.versions
                    ).match()
                }
            )
        }
    }

    test("Undibacterium_unclassified - fasta - bam out - filter") {

        when {

            params {
                minimap2_args      = "-x map-hifi"
                samtools_view_args = "-F 0x904"
                samtools_sort_args = "--write-index -l1"
            }

            process {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/genomic_data/baUndUnlc1/pacbio/fasta/HiFi.reads.fasta', checkIfExists: true),
                    0,
                    [0, 500],
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/assembly/draft/baUndUnlc1.hic.hap1.p_ctg.fa.gz', checkIfExists: true)
                ]).combine(FASTXALIGN_PYFASTXINDEXFASTA.out.index, by: 0)
                .map { meta, reads, chunkn, range, reference, index, count ->
                    [ meta, reads, index, chunkn, range, reference ]
                }
                input[1] = true
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    bam(process.out.bam.get(0).get(1)).getStatistics(),
                    process.out.versions
                    ).match()
                }
            )
        }
    }

    test("Undibacterium_unclassified - fasta - paf out") {

        when {

            params {
                minimap2_args      = "-x map-hifi"
                samtools_view_args = ""
                samtools_sort_args = ""
            }

            process {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/genomic_data/baUndUnlc1/pacbio/fasta/HiFi.reads.fasta', checkIfExists: true),
                    0,
                    [0, 500],
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/assembly/draft/baUndUnlc1.hic.hap1.p_ctg.fa.gz', checkIfExists: true)
                ]).combine(FASTXALIGN_PYFASTXINDEXFASTA.out.index, by: 0)
                .map { meta, reads, chunkn, range, reference, index, count ->
                    [ meta, reads, index, chunkn, range, reference ]
                }
                input[1] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("sarscov2 - bam - stub") {

        options "-stub"

        when {

            params {
                minimap2_args      = ""
                samtools_view_args = ""
                samtools_sort_args = ""
            }

            process {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/genomic_data/baUndUnlc1/pacbio/fasta/HiFi.reads.fasta', checkIfExists: true),
                    0,
                    [0, 500],
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/assembly/draft/baUndUnlc1.hic.hap1.p_ctg.fa.gz', checkIfExists: true)
                ]).combine(FASTXALIGN_PYFASTXINDEXFASTA.out.index, by: 0)
                .map { meta, reads, chunkn, range, reference, index, count ->
                    [ meta, reads, index, chunkn, range, reference ]
                }
                input[1] = true
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
