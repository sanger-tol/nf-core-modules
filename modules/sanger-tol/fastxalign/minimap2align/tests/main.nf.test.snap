{
    "Undibacterium_unclassified - fasta - bam out - filter": {
        "content": [
            {
                "maxReadLength": 26308,
                "minReadLength": 1368,
                "meanReadLength": 8157,
                "maxQuality": 60,
                "minQuality": 20,
                "meanQuality": 59,
                "readCount": 428,
                "duplicateReadCount": 0,
                "sorted": false
            },
            [
                "versions.yml:md5,3c11d63382002b700419986fc18f8faa"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-20T11:43:13.559285"
    },
    "sarscov2 - bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "HiFi.reads.fasta.0.test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "HiFi.reads.fasta.0.test.paf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,3c11d63382002b700419986fc18f8faa"
                ],
                "bam": [
                    [
                        {
                            "id": "test"
                        },
                        "HiFi.reads.fasta.0.test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "paf": [
                    [
                        {
                            "id": "test"
                        },
                        "HiFi.reads.fasta.0.test.paf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3c11d63382002b700419986fc18f8faa"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-20T11:44:04.167447"
    },
    "Undibacterium_unclassified - fasta - bam out - no filter": {
        "content": [
            {
                "maxReadLength": 26308,
                "minReadLength": 1368,
                "meanReadLength": 8121,
                "maxQuality": 60,
                "minQuality": 0,
                "meanQuality": 50,
                "readCount": 501,
                "duplicateReadCount": 0,
                "sorted": false
            },
            [
                "versions.yml:md5,3c11d63382002b700419986fc18f8faa"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-20T11:42:48.915403"
    },
    "Undibacterium_unclassified - fasta - paf out": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,3c11d63382002b700419986fc18f8faa"
                ],
                "bam": [
                    
                ],
                "paf": [
                    
                ],
                "versions": [
                    "versions.yml:md5,3c11d63382002b700419986fc18f8faa"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-20T11:43:39.452366"
    }
}